[SPARK-35455][SQL] Unify empty relation optimization between normal and AQE optimizer

ulysses-you opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error message, please read the guideline first:
     https://spark.apache.org/error-message-guidelines.html
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->

* remove `EliminateUnnecessaryJoin`, using `AQEPropagateEmptyRelation` instead.
* eliminate join, aggregate, limit, repartition, sort, generate which is beneficial.

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
Make `EliminateUnnecessaryJoin` available with more case.

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->
No.

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->
Add test.

ulysses-you committed:
Enhance EliminateUnnecessaryJoin

ulysses-you committed:
fix

SparkQA commented:
**[Test build #138730 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138730/testReport)** for PR 32602 at commit [`5097247`](https://github.com/apache/spark/commit/509724757f59e9b01ccb0774fb6d1cc56e75f9dc).

ulysses-you commented:
cc @maropu @cloud-fan @c21

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/EliminateUnnecessaryJoin.scala:
```diff
@@ -59,19 +66,33 @@ object EliminateUnnecessaryJoin extends Rule[LogicalPlan] {
         case _ => false
       }
+
+    case LocalRelation(_, data, isStreaming) if !isStreaming =>
```
if we want to handle `LocalRelation`, then it's not AQE specific and we can do it in the normal optimizer?

ulysses-you replied:
yes, but currently we have no chance to do normal optimizer at AQE side. Maybe we can let some rules which in `Optimizer` also available at `AQEOptimizer` in future ?

cloud-fan replied:
Since AQE is on by default, it's not a big issue but more about code cleanness. How about this:
1. `EliminateUnnecessaryJoin` should only deal with `LocalRelation`, and appears in both the normal optimizer and AQE optimizer
2. AQE optimizer adds a new rule to turn empty query stage into empty `LocalRelation`

ulysses-you replied:
sgtm, and just find a exists rule `PropagateEmptyRelation`. 

How about this updating ?
* Add a rule `ConvertToLocalRelation` to  turn empty query stage into empty LocalRelation
* Make `PropagateEmptyRelation` appears in AQE optimizer
* Reduce `EliminateUnnecessaryJoin` to only handle naaj/semi/anti joins

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43252/

ulysses-you committed:
LocalRelation early

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43252/

ulysses-you committed:
fix

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43252/

SparkQA commented:
**[Test build #138736 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138736/testReport)** for PR 32602 at commit [`165077b`](https://github.com/apache/spark/commit/165077b63b99df3ed20634190ae4df8dd7895ea8).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43258/

SparkQA commented:
**[Test build #138730 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138730/testReport)** for PR 32602 at commit [`5097247`](https://github.com/apache/spark/commit/509724757f59e9b01ccb0774fb6d1cc56e75f9dc).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43258/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43258/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/138730/

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/ConvertToLocalRelation.scala:
```diff
@@ -0,0 +1,32 @@
+ */
+object ConvertToLocalRelation extends Rule[LogicalPlan] {
+  override def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case l @ LogicalQueryStage(_, stage: QueryStageExec) if stage.resultOption.get().isDefined &&
```
Sorry I made the wrong decision. This may change the output partitioning and is not always safe/beneficial (we may add extra shuffles in the planning phase later).

Looking at the optimizations for empty local relations, some of them are likely beneficial and we can always do: eliminate join, aggregate, limit, repartition, sort, generate

Some may not that beneficial and we shouldn't do: simplify union, eliminate project/filter/sample.

My new idea:
1. Create a new rule `PropagateEmptyRelationBasic`, which deals with local relaion only, and is not very beneficial (eliminate project, filter, etc), and runs in the normal optimizer only
2. Create a new rule `PropagateEmptyRelationAdvanced`, which deals with both local relation and query stage, and is very beneficial (eliminate join, aggregate, etc.), and runs in both normal and AQE optimizer

The old `EliminateUnnecessaryJoin` and `PropagateEmptyRelation` rules should be removed and merged into the new rules.

`PropagateEmptyRelationAdvanced` may not be able to access QueryStageExec which is in sql/core. We can let this rule take `checkRowCount` functiton as a parameter.

ulysses-you replied:
I see the issue. If we worry about the `LocalRelation` output partitioning, we can just mark `LocalRelationExec` output partitioning as `SinglePartition` to avoid extra shuffle. But it doesn't work with optimization like non empty left semi/anti elimintaion.

> we can always do: eliminate join, aggregate

Not sure we can always do this if we don't want to introduce extra shuffle. And the issue it has already existed in current `EliminateUnnecessaryJoin`, like this plan
```
Aggregate (same key with join)
  Join Inner
    LocalRelation
    xxx
```

An another idea, if we plan to support extra shuffle later and don't expect introduce shuffle at AQE optimzier side, then is it better to check the physical plan `requiredChildDistribution` ? We can only allow one node which has a valid `requiredChildDistribution` (not `UnspecifiedDistribution`) in one query stage, and skip optimize if one query stage has two or more valid `requiredChildDistribution` nodes. Thus we can run `PropagateEmptyRelation` safely.

cloud-fan replied:
If we turn a broadcast stage into local relation without changing other plan parts, seems we will broadcast the local relation again. `SinglePartition` can't help here.

ulysses-you replied:
If it's hard to avoid introduce shuffle at AQE optimizer, how about add extra shuffle check between AQE optimizer and stage preparation ? Then it won't affect the extra shuffle in stage preparation.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/138736/

ulysses-you committed:
split PropagateEmptyRelation

ulysses-you committed:
fix

SparkQA commented:
**[Test build #138793 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138793/testReport)** for PR 32602 at commit [`7b80db0`](https://github.com/apache/spark/commit/7b80db02cc0a07a48aa378e03658d5c6024459ba).

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PropagateEmptyRelation.scala:
```diff
@@ -72,11 +65,72 @@ object PropagateEmptyRelation extends Rule[LogicalPlan] with PredicateHelper wit
+ *     - Aggregate with all empty children and at least one grouping expression.
+ *     - Generate(Explode) with all empty children. Others like Hive UDTF may return results.
+ *
+ * @param checkRowCount At AQE side, we use the query stage stats to check the check.
```
`At AQE side, we use this function to check if a plan has output rows or not`

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PropagateEmptyRelation.scala:
```diff
@@ -72,11 +65,72 @@ object PropagateEmptyRelation extends Rule[LogicalPlan] with PredicateHelper wit
+    plan.output.map{ a => Alias(cast(Literal(null), a.dataType), a.name)(a.exprId) }
+
+  // We can not use transformUpWithPruning here since this rule is used by both normal Optimizer
+  // and AQE Optimizer. And this may only effective at AQE side.
```
ah good point. I think there is a way to overcome it:
1. Create an abstract class `PropagateEmptyRelationBase` that contains util functions and optimizes expensive operators such as join, aggregate, etc.
2. Create a rule `PropagateEmptyRelation extends PropagateEmptyRelationBase` that additionally optimzes project, filter, etc.
3. Create a rule `AQEPropagateEmptyRelation extends PropagateEmptyRelationBase` that overrides some util functions like `isEmptyPlan`.

Then this two rules can define their transformation prunning separatedly.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43316/

ulysses-you committed:
split PropagateEmptyRelationBase

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43316/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43316/

ulysses-you committed:
fix

ulysses-you commented on sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AQEPropagateEmptyRelation.scala:
```diff
@@ -0,0 +1,53 @@
+  }
+
+  // TODO we need use transformUpWithPruning instead of transformUp
+  def apply(plan: LogicalPlan): LogicalPlan = plan.transformUp {
```
@cloud-fan , if we want to use `transformUpWithPruning ` at AQE optimizer side, we need to some more work like add pattern at `LogicalQueryStage`. So this PR does not do the change, just use `transformUp`. Do you think it's OK ?

cloud-fan replied:
yea it's OK as it's not a regression. cc @sigmod

ulysses-you replied:
@cloud-fan seems we don't need to use `transformUpWithPruning` here since the AQE Optimizer always run once rather than fixed point ?

SparkQA commented:
**[Test build #138803 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138803/testReport)** for PR 32602 at commit [`48dd92a`](https://github.com/apache/spark/commit/48dd92a741ab87c50e7907d9169148994bf62a3a).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43325/

SparkQA commented:
**[Test build #138793 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138793/testReport)** for PR 32602 at commit [`7b80db0`](https://github.com/apache/spark/commit/7b80db02cc0a07a48aa378e03658d5c6024459ba).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/138793/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43325/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43325/

SparkQA commented:
**[Test build #138803 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138803/testReport)** for PR 32602 at commit [`48dd92a`](https://github.com/apache/spark/commit/48dd92a741ab87c50e7907d9169148994bf62a3a).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `abstract class PropagateEmptyRelationBase extends Rule[LogicalPlan] with CastSupport `

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/138803/

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PropagateEmptyRelation.scala:
```diff
@@ -137,3 +192,10 @@ object PropagateEmptyRelation extends Rule[LogicalPlan] with PredicateHelper wit
   }
 }
+
+object PropagateEmptyRelation extends PropagateEmptyRelationBase {
```
can we merge PropagateEmptyRelationBasic into this rule?

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AQEPropagateEmptyRelation.scala:
```diff
@@ -0,0 +1,53 @@
+object AQEPropagateEmptyRelation extends PropagateEmptyRelationBase {
+
+  override protected def checkRowCount(plan: LogicalPlan, hasRow: Boolean): Option[Boolean] = {
+    Some(plan match {
```
Can we revert to the previous version? Returning None means we don't know if it's empty or not and should not do optimization based on it.

cloud-fan replied:
https://github.com/apache/spark/pull/32602/files#diff-b4320e1fb79936232c73e55f34b8a14486b0351a435eeb4914d81a8a756b6afbL58

ulysses-you committed:
merge

ulysses-you committed:
nit

ulysses-you commented:
thank you @cloud-fan , addressed the comments:
* merge `PropagateEmptyRelation` so now at normal optimizer there is no change
* revert the two method that used by AQE side, return `None` as default

SparkQA commented:
**[Test build #138848 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138848/testReport)** for PR 32602 at commit [`1220087`](https://github.com/apache/spark/commit/12200876b863f10b428741e0ac69ee76d5802b1d).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43370/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43370/

SparkQA commented:
**[Test build #138848 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138848/testReport)** for PR 32602 at commit [`1220087`](https://github.com/apache/spark/commit/12200876b863f10b428741e0ac69ee76d5802b1d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

ulysses-you committed:
fix

SparkQA commented:
**[Test build #138854 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138854/testReport)** for PR 32602 at commit [`e26df96`](https://github.com/apache/spark/commit/e26df96b72806ace84c72e1f134278e2b20b4457).

SparkQA commented:
**[Test build #138854 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138854/testReport)** for PR 32602 at commit [`e26df96`](https://github.com/apache/spark/commit/e26df96b72806ace84c72e1f134278e2b20b4457).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43376/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/138854/

ulysses-you committed:
fix

SparkQA commented:
**[Test build #138855 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138855/testReport)** for PR 32602 at commit [`f7a14cf`](https://github.com/apache/spark/commit/f7a14cf49ac27336195dbcb6285b2b2e5f96df38).

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43376/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43377/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43376/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43377/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43377/

SparkQA commented:
**[Test build #138855 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138855/testReport)** for PR 32602 at commit [`f7a14cf`](https://github.com/apache/spark/commit/f7a14cf49ac27336195dbcb6285b2b2e5f96df38).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/138855/

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PropagateEmptyRelation.scala:
```diff
@@ -20,65 +20,73 @@ package org.apache.spark.sql.catalyst.optimizer
-object PropagateEmptyRelation extends Rule[LogicalPlan] with PredicateHelper with CastSupport {
-  private def isEmptyLocalRelation(plan: LogicalPlan): Boolean = plan match {
+abstract class PropagateEmptyRelationBase extends Rule[LogicalPlan] with CastSupport {
+  /**
```
how about
```
protected def isEmpty(plan): Boolean = plan match {
  case p: LocalRelation => p.data.isEmpty
  case _ => false
}

protected def nonEmpty(plan): Boolean = plan match {
  case p: LocalRelation => p.data.nonEmpty
  case _ => false
}
```

in AQE rule
```
override def isEmpty(plan): Boolean = super.isEmpty(plan) || getRowCount(plan).exists(_ == 0)

override def nonEmpty(plan): Boolean = super.nonEmpty(plan) || getRowCount(plan).exists(_ > 0)

private def getRowCount(plan): Option[Int] = plan match {
  case LogicalQueryStage(_, stage: QueryStageExec) if stage.resultOption.get().isDefined => 
    stage.getRuntimeStatistics.rowCount
  case _ => None
}
```

ulysses-you replied:
yes, looks more cleaner

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PropagateEmptyRelation.scala:
```diff
@@ -20,65 +20,73 @@ package org.apache.spark.sql.catalyst.optimizer
-          Project(outputAliases, newPlan)
-        }
-      }
+  protected def propagateEmptyRelationAdvanced: PartialFunction[LogicalPlan, LogicalPlan] = {
```
nit: `commonApplyFunc`

cloud-fan replied:
`single column NULL-aware anti join (NAAJ)` is AQE only, I think it's better to match it only in the AQE rule

ulysses-you replied:
Extracted this pattern to AQE side.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PropagateEmptyRelation.scala:
```diff
@@ -137,3 +149,53 @@ object PropagateEmptyRelation extends Rule[LogicalPlan] with PredicateHelper wit
+ *    - Project/Filter/Sample with all empty children.
+ */
+object PropagateEmptyRelation extends PropagateEmptyRelationBase {
+  private def propagateEmptyRelationBasic: PartialFunction[LogicalPlan, LogicalPlan] = {
```
nit: `applyFunc`

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PropagateEmptyRelation.scala:
```diff
@@ -137,3 +149,53 @@ object PropagateEmptyRelation extends Rule[LogicalPlan] with PredicateHelper wit
 }
+
+/**
+ * Rule [[PropagateEmptyRelationBase]] at normal optimizer side.
```
let's also mention why we don't do it in the AQE side: these optimizations are not signigicant and may introduce extra exchanges.

ulysses-you replied:
Updated.

ulysses-you committed:
empty

ulysses-you committed:
naaj

ulysses-you committed:
comment

ulysses-you committed:
nit

ulysses-you committed:
inline

SparkQA commented:
**[Test build #138867 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138867/testReport)** for PR 32602 at commit [`767dd92`](https://github.com/apache/spark/commit/767dd92618461c2891e6aeb73d99f320f1601f17).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43389/

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PropagateEmptyRelation.scala:
```diff
@@ -26,59 +26,48 @@ import org.apache.spark.sql.catalyst.rules._
+ * The rule used by both normal Optimizer and AQE Optimizer for:
+ *  1. Binary-node Logical Plans
+ *     - Join with one or two empty children (including Intersect/Except).
+ *     - Join is single column NULL-aware anti join (NAAJ)
```
we should remove this now.

ulysses-you replied:
Moved it to `AQEPropagateEmptyRelation`

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PropagateEmptyRelation.scala:
```diff
@@ -26,59 +26,48 @@ import org.apache.spark.sql.catalyst.rules._
- *    - Join with false condition.
- *    - Aggregate with all empty children and at least one grouping expression.
- *    - Generate(Explode) with all empty children. Others like Hive UDTF may return results.
+ * The rule used by both normal Optimizer and AQE Optimizer for:
```
```
The base class of two rules in the normal and AQE Optimizer. It simplifies query plans with
empty or non-empty relations:
  ...
```

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PropagateEmptyRelation.scala:
```diff
@@ -26,59 +26,48 @@ import org.apache.spark.sql.catalyst.rules._
     // as stateful streaming joins need to perform other state management operations other than
     // just processing the input data.
     case p @ Join(_, _, joinType, conditionOpt, _)
-        if !p.children.exists(_.isStreaming) =>
```
the previous indentation is correct.

ulysses-you replied:
reverted

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PropagateEmptyRelation.scala:
```diff
@@ -137,3 +127,54 @@ object PropagateEmptyRelation extends Rule[LogicalPlan] with PredicateHelper wit
 }
+
+/**
+ * Rule [[PropagateEmptyRelationBase]] at normal optimizer side with the extra pattern:
```
```
This rule runs in the normal optimizer and optimizes more cases
compared to [[PropagateEmptyRelationBase]]:
  ...
```

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AQEOptimizer.scala:
```diff
@@ -27,7 +28,9 @@ import org.apache.spark.util.Utils
 class AQEOptimizer(conf: SQLConf) extends RuleExecutor[LogicalPlan] {
   private val defaultBatches = Seq(
-    Batch("Eliminate Unnecessary Join", Once, EliminateUnnecessaryJoin),
+    Batch("Propagate Empty LocalRelation", Once,
```
`LocalRelation` -> `Relations`?

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:
```diff
@@ -1307,6 +1308,69 @@ class AdaptiveQueryExecSuite
+    withSQLConf(SQLConf.ADAPTIVE_EXECUTION_ENABLED.key -> "true",
+      SQLConf.AUTO_BROADCASTJOIN_THRESHOLD.key -> "-1") {
+      Seq(
+        // left semi join and empty left side
```
We can't optimize this before this PR?

ulysses-you replied:
yeah, we cann't. Before we only check right side with `LeftSemi/LeftAnti`.

And the test should use different column to do filter and join  in case of `InferFiltersFromConstraints` make right side empty.  Updated it.

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43389/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43389/

ulysses-you committed:
test

ulysses-you committed:
comment

ulysses-you committed:
Relations

ulysses-you committed:
indentation

SparkQA commented:
**[Test build #138870 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138870/testReport)** for PR 32602 at commit [`0754936`](https://github.com/apache/spark/commit/0754936c6a053f3f8b6d7a440fca398621c6eb21).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43392/

SparkQA commented:
**[Test build #138867 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138867/testReport)** for PR 32602 at commit [`767dd92`](https://github.com/apache/spark/commit/767dd92618461c2891e6aeb73d99f320f1601f17).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43392/

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:
```diff
@@ -1307,6 +1308,69 @@ class AdaptiveQueryExecSuite
     }
   }
 
+  test("SPARK-35455: Enhance EliminateUnnecessaryJoin - single join") {
```
let's update the test name and PR title: `Unify empty relation optimization between normal and AQE optimizer`

ulysses-you replied:
Updated it and also updated the PR title.

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:
```diff
@@ -1307,6 +1308,69 @@ class AdaptiveQueryExecSuite
+    }
+  }
+
+  test("SPARK-35455: Enhance EliminateUnnecessaryJoin - multi join") {
```
ditto

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43392/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/138867/

SparkQA commented:
**[Test build #138870 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138870/testReport)** for PR 32602 at commit [`0754936`](https://github.com/apache/spark/commit/0754936c6a053f3f8b6d7a440fca398621c6eb21).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/138870/

ulysses-you committed:
test name

ulysses-you committed:
nit

SparkQA commented:
**[Test build #138899 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138899/testReport)** for PR 32602 at commit [`624e45e`](https://github.com/apache/spark/commit/624e45e27513933e9fe5615601d0dddac341a258).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43421/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43421/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43421/

c21 reviewed (approved):
LGTM

SparkQA commented:
**[Test build #138899 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138899/testReport)** for PR 32602 at commit [`624e45e`](https://github.com/apache/spark/commit/624e45e27513933e9fe5615601d0dddac341a258).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/138899/

cloud-fan commented:
thanks, merging to master!

ulysses-you commented:
@cloud-fan thank you for the discussion and merging ! @c21 thank you for the review !

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AQEOptimizer.scala:
```diff
@@ -27,7 +28,9 @@ import org.apache.spark.util.Utils
-    Batch("Eliminate Unnecessary Join", Once, EliminateUnnecessaryJoin),
+    Batch("Propagate Empty Relations", Once,
+      AQEPropagateEmptyRelation,
+      UpdateAttributeNullability),
```
shall we also put `ConvertToLocalRelation` in this batch?

ulysses-you replied:
yeah, I noticed it. We can put it so that we can propagate empty through `project/filter`. like such case:
```
Aggregate
  Project
    Join
      Shuffle
```
But it need to isolate normal and AQE due to `transformWithPruning`.

Otherhand I feel that it's similar if we just let `AQEPropagateEmptyRelation` support propagate `project/filter`. and the later is simpler.

cloud-fan replied:
It's a bit different:
```
Project
  Shuffle Stage
```
For the above case, we don't want to optimize it as the benefit is too small

```
Project
  Sort
    Shuffle Stage
```
For the above case, we will optimize Sort -> Shuffle Stage to empty relation first. Then it makes sense to optimize further and optimize out project, as the shuffle stage is already gone.

So adding `ConvertToLocalRelation` looks the best solution here.

ulysses-you replied:
ah I see, will do this.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PropagateEmptyRelation.scala:
```diff
@@ -137,3 +125,55 @@ object PropagateEmptyRelation extends Rule[LogicalPlan] with PredicateHelper wit
+ *    - Project/Filter/Sample with all empty children.
+ *
+ * The reason why we don't apply this rule at AQE optimizer side is: the benefit is not big enough
+ * and it may introduce extra exchanges.
```
After more thought, I think this is a big performance issue if we can't propagate empty relations through project/filter which are quite common. The risk of introducing new shuffles is relatively small compared to this.

@ulysses-you can we move all the logic to `PropagateEmptyRelationBase`? `PropagateEmptyRelation` should not have any extra logic.
